jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9PictureMap.kt

```diff
@@ -0,0 +1,332 @@
+        return c.item
+    }
+
+    /** Get the latest picture in the tracker.  */
+    val latestPicture: Vp9Picture?
+        get() = getIndex(lastIndex)
+
+    fun insert(pictureId: Int, picture: Vp9Picture): Boolean {
+        val index = indexTracker.update(pictureId)
+        val ret = super.insertItem(picture, index)
+        if (ret) {
+            numCached++
+            if (firstIndex == -1 || index < firstIndex) {
+                firstIndex = index
+            }
+        }
+        return ret
+    }
+
+    /**
+     * Called when an item in the cache is replaced/discarded.
+     */
+    override fun discardItem(item: Vp9Picture) {
+        numCached--
+    }
+
+    fun findBefore(frame: Vp9Frame, pred: (Vp9Frame) -> Boolean): Vp9Frame? {
+        val lastIndex = lastIndex
+        if (lastIndex == -1) {
+            return null
+        }
+        val index = indexTracker.interpret(frame.pictureId)
+        val searchStartIndex = Integer.min(index, lastIndex)
+        val searchEndIndex = Integer.max(lastIndex - size, firstIndex - 1)
+        return doFind(
+            pred = pred,
+            startIndex = searchStartIndex,
+            endIndex = searchEndIndex,
+            startLayer = frame.spatialLayer - 1,
+            increment = -1
+        )
+    }
+
+    fun findAfter(frame: Vp9Frame, pred: (Vp9Frame) -> Boolean): Vp9Frame? {
+        val lastIndex = lastIndex
+        if (lastIndex == -1) {
+            return null
+        }
+        val index = indexTracker.interpret(frame.pictureId)
+        if (index > lastIndex) {
+            return null
+        }
+        val searchStartIndex = Integer.max(index, Integer.max(lastIndex - size + 1, firstIndex))
+        return doFind(
+            pred = pred,
+            startIndex = searchStartIndex,
+            endIndex = lastIndex + 1,
+            startLayer = frame.spatialLayer + 1,
+            increment = 1
+        )
+    }
+
+    private fun doFind(
+        pred: (Vp9Frame) -> Boolean,
+        startIndex: Int,
+        endIndex: Int,
+        startLayer: Int,
+        increment: Int
+    ): Vp9Frame? {
+        var index = startIndex
+        var layer = startLayer
+        var firstPicture = true
+        while (index != endIndex) {
+            val picture = getIndex(index)
+            if (picture != null) {
+                if (!firstPicture) {
+                    layer = if (increment < 0) picture.frames.size - 1 else 0
+                }
+                while (layer in picture.frames.indices) {
+                    val frame = picture.frames[layer]
+                    if (frame != null && pred(frame)) {
+                        return frame
+                    }
+                    layer += increment
+                }
+            }
+            index += increment
+            firstPicture = false
+        }
+        return null
+    }
+
+    /** Like Rfc3711IndexTracker, but for picture IDs (so with a rollover
```

Comment:
Can we factor out the shared functionality?

Reply:
I'll take a look and see how easy it would be to do.

Certainly I should look at sharing code between the VP8 and VP9 code.